src/main/java/com/google/devtools/build/lib/worker/WorkerExecRoot.java

```diff
@@ -99,16 +99,16 @@ private void populateInputsAndDirsToCreate(
       }
     }
 
-    // Add all ouput directories.
-    dirsToCreate.addAll(outputs.dirs());
-
     // And all ancestor directories of outputs. Note that we don't add the files themselves -- any
     // pre-existing files that have the same path as an output should get deleted.
     for (PathFragment path : Iterables.concat(outputs.files(), outputs.dirs())) {
       for (int i = 0; i < path.segmentCount(); i++) {
         dirsToCreate.add(path.subFragment(0, i));
       }
     }
+
+    // Add all ouput directories.
+    dirsToCreate.addAll(outputs.dirs());
```

Comment:
I think there should be a comment here somewhere about the order of set additions mattering, and maybe (I know some people hate it) change the method signature to say LinkedHashSet. I certainly had to look up the caller to double-check that it's a LinkedHashSet. Otherwise, this wouldn't work.

Reply:
Good point. Done.

Reply:
I understand the desire to change the type of the variable to clarify intent, but this change in the variable declaration is a no-op: a future reader of this code (or maybe even a reformatter!) won't understand why this variable looks different than others and may undo the "fix".

I think I suggested this in the original review for this code and I think the author agreed to it... but then I'm not sure what happened with the suggestion. What about changing this (and the other variables) to a `ImmutableSortedSet.Builder`, passing the builder around to all the functions that append to it, and then constructing the final set here?

This would clearly show that this is passed to functions as an in/out parameter, and would also show that order is important (vs. being a hashset, which is only an implementation detail of an ordered container).

Reply:
@jmmv pointed out that because `ImmutableSortedSet` extends `ImmutableSet`, using `ImmutableSortedSet.Builder` may not necessarily fix the potential future refactoring breakage problem. I suppose we should merge the PR in its current state, and think of how to improve readability in such cases as a followup.

Reply:
@tetromino @jmmv ya - let me know how you want me to proceed, I'm happy to make changes. I noticed that ImmutableSortedSet.builder() is deprecated and didn't see an API that would preserve insertion order though. Maybe I missed something?

Reply:
I think your original idea to use a `LinkedHashSet` was the right way to go.

An `ImmutableSortedSet.Builder` sorts entries via a comparator (the default natural order comparator, if you use the deprecated `ImmutableSortedSet.builder()` or the new `ImmutableSortedSet.naturalOrder()`), not by insertion order.

For directories that would incidentally work, since a subdirectory would be sorted after its parent directory, but a data structure that preserves insertion order would more explicitly fit your intent and in any case would be more efficient.

Unfortunately, I don't see an immutable insertion order preserving set in Guava.